### DISCUSSION TITLE
What was the original reason on pulling in `rodio`?

### DISCUSSION THREAD

**Discussion by hasezoey** [tramhao/termusic#203] [Q&A] [answered]
I just noticed that `playback::rusty_backend` is [`rodio`](https://github.com/RustAudio/rodio/) but pulled into the project.
What was the original reason it has been pulled-in and should it ever be swapped back?

**Comment by hasezoey** [comment]
something related i came across:

> Previously I used rodio, but rodio doesn't support seek

https://github.com/tramhao/termusic/issues/33#issuecomment-1003690797

so if the current state of `rodio` is anything to go be, it still does not support seeking https://github.com/RustAudio/rodio/issues/176

**Comment by tramhao** [comment]
Yes, besides that, I also need to modify some  signal for gapless playback.

**Comment by yara-blue** [comment]
As of april this year Rodio supports seeking, many other features have been added since, such as automatic gain control. 

Please let us know what is missing from rodio that you need. We are currently redesigning much of the API and could use your feedback.

I hope you will be able to depend on rodio directly instead of the current vendoring situation (copying in the code) in the near future. The current vendoring situation is rather problematic from a legal point of view. 

Rodio code is licensed under both the MIT and Apache License, both require attribution. That means you should mention which parts of your source code has been copied from Rodio.

**Reply by hasezoey** [reply]
Thanks for your update regarding seeking in rodio, i have tried to switch over to it and am missing some features, or at least a good way without some not-too-great workarounds:
- send a message on EOS of a source [(see this code in our implementation)](https://github.com/tramhao/termusic/blob/01c252d24dc04f30eb83e3a6047e3137de548352/playback/src/rusty_backend/sink.rs#L298-L316) (suggestion would be to get the handle to the `sleep_until_end` directly, or a function which accepts a callback and is similar to `source::Done`)
- possibly a way to add / replace other source modifiers after all the other processing in `Sink::append` [(see this code in our implementation)](https://github.com/tramhao/termusic/blob/01c252d24dc04f30eb83e3a6047e3137de548352/playback/src/rusty_backend/sink.rs#L133-L151) (would be necessary to replace `speed` with our own `tempo_stretch`; or implement it in rodio directly)
- there is likely a possible de-sync between [the track_position getting for events](https://github.com/tramhao/termusic/compare/master...hasezoey:termusic:rodio#diff-808b582317c038c19a268d4dcced34210b24948aa6047c43995fdc9670cc6b5cR258-R270) and the `Sink::get_pos` which is after `speed` / `delay`

[For reference, this is the switchover i tried](https://github.com/tramhao/termusic/compare/master...hasezoey:termusic:8c151ec15affede47cc5d6c6d910dc1166b0f651), it can play music but it does not fully work as before yet.
We could likely implement a workaround to detect track / queue changes, but it would be a lot better if we could just do a `sleep_until_end` in another thread or something similar to `source::Done` which sends a message over a channel or accepts a callback similar to `periodic_access`.

Though maybe i am just missing some things rodio provides.

Do you maybe have some suggestions on where we are using rodio wrongly / better ways to do what we are currently doing? (in the context of the switch back to upstream)

---

> I hope you will be able to depend on rodio directly instead of the current vendoring situation (copying in the code) in the near future. The current vendoring situation is rather problematic from a legal point of view.
> Rodio code is licensed under both the MIT and Apache License, both require attribution. That means you should mention which parts of your source code has been copied from Rodio.

this is my first time dealing with such a problem, what would be required for this to be resolved in the current copied state?
- a simple header in each file (SPDX?)
- a attribution / license file in a root directly (would restructure to have only rodio related code there)
- both licenses or is one (MIT or Apache) enough?

**Reply by hasezoey** [reply]
seeing as @dvdsk did not link it here (and github not showing external referencing in discussions), a new issue in rodio was created to address those comments also searchable for other rodio users: https://github.com/RustAudio/rodio/issues/637

**Reply by yara-blue** [reply]
As you have already seen I moved my answers to the rodio api specific questions to the rodio repo (and the had dinner before I posted this).

Regarding the license issue:

> this is my first time dealing with such a problem, what would be required for this to be resolved in the current copied state?

I do not think there is any rush, as long as your willing to eventually fix it I'm happy, I do not know about the rest of the team though. If you choose to keep the code vendor-ed any of the solution you propose is good. Another alternative would be to depend on a modified fork of rodio. But that would introduce a git dependency in your app, something you can not have if you release on crates.io.

> both licenses or is one (MIT or Apache) enough?

from the rodio repo:
> Licensed under either of
> 
>     Apache License, Version 2.0, ([LICENSE-APACHE](https://github.com/RustAudio/rodio/blob/master/LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0), or
>     MIT license ([LICENSE-MIT](https://github.com/RustAudio/rodio/blob/master/LICENSE-MIT) or http://opensource.org/licenses/MIT)
> 
> at your option.

So you can pick whatever fits best.

As  I side note,  I see you have some GPLv3 dependencies. Those require that your project becomes GPLv3 too. Its no a big deal GPLv3 is stricter then MIT or Apache. You could see GPLv3 as an extension of MIT/Apache. See more about the GPL here: https://www.gnu.org/licenses/quick-guide-gplv3.html

**Reply by hasezoey** [reply]
> If you choose to keep the code vendor-ed any of the solution you propose is good. Another alternative would be to depend on a modified fork of rodio

A modified fork would also be my go-to option, but i am working with existing code.

As for how i plan on resolving this, i think about just throwing out most of the vendor-ed code, only keeping / re-implementing the parts that are necessary to be modified (like as discussed a custom `Sink`) (with proper attribution where necessary).

> As I side note, I see you have some GPLv3 dependencies. Those require that your project becomes GPLv3 too. Its no a big deal GPLv3 is stricter then MIT or Apache. You could see GPLv3 as an extension of MIT/Apache. See more about the GPL here: https://www.gnu.org/licenses/quick-guide-gplv3.html

I also thought this *might* be a conflict, but there too i am working with existing code, i also am planning on completely re-writing those sections to be more generic / dynamic. (though i dont know if we could just re-license that code as it has quite heavily diverged already)

> Those require that your project becomes GPLv3 too

Personally, i would also have preferred that some of this code (`lib/src/songtag/*`) would be a separate crate, but it is what it is currently. (would that even be enough?)
Or were you referencing something else?

**Reply by yara-blue** [reply]
> (would that even be enough?)

No, it does not matter if its a separate crate or copied in, if any GPL code is compiled into your binary your project is now GPL. The idea is that a company can not just copy an open source project, improve it and then make it closed source. Given termusic is not for profit I do not think it should be a problem to make termusic GPL licensed. If you use a GPL project to generate a program it might be different, for example gcc the c compiler is GPL but not every program compiled using gcc has to be GPL. You are allowed to use GPL for commercial proposes. But you are not allowed build a GUI for gcc and sell that.

**Reply by yara-blue** [reply]
> only keeping / re-implementing the parts that are necessary to be modified (like as discussed a custom `Sink`) (with proper attribution where necessary).

If you write it its your code and does not need any attribution. So a custom `Sink` made out of rodio parts you use from the rodio crate needs no attribution. Only when you copy over the code.

(Technically every dependency should be credited somewhere in the app (something like an about page) but I only see huge projects ever do that).

**Reply by hasezoey** [reply]
I have now created #389 which switches to upstream rodio except for a custom `Sink`, which has attribution as it is still basically a copy of the upstream one but somewhat modified.

Are those changes enough to resolve the rodio license situation @dvdsk?

**Reply by yara-blue** [reply]
Thats perfect, thank you very much for making those changes! Oh and by the way the latest rodio has automatic gain control which could be nice to have in termusic.

**Reply by hasezoey** [reply]
> Oh and by the way the latest rodio has automatic gain control which could be nice to have in termusic.

Sound good, though it had not been implemented in any other backend, so it would likely take a little more effort.

I am not too much into audio and have not tried it yet, so i dont quite understand what / why it would be needed. Is this something similar to ReplayGain but for live sources and adjusted over time?
Edit: created #390

**Reply by yara-blue** [reply]
> I am not too much into audio

Same tbh, I wanted to build a podcast app and I have ended up maintaining rodio and writing a streaming/download manager library :sweat_smile:. Maybe I'll get on that partially finished podcast app once thats done.

> Is this something similar to ReplayGain but for live sources and adjusted over time?

Yes, it allows you to limit the maximum volume and keep the average volume around a set limit. Can be useful if a song has very loud and very quite bits.

**Reply by hasezoey** [reply, marked as answer]
For completeness, #389, the PR to switch back to upstream `rodio` has been merged.
